### ISSUE TITLE
An implementation of display-resample that works in tandem with VRR displays?

### ISSUE THREAD

**Issue by dokur0chan** [mpv-player/mpv#12005]
I want to preface this by saying that my understanding of this could be fundamentally flawed, so I wanted to ask this to understand why this couldn't be feasible in mpv.

As of right now, `video-sync=display-resample` and VRR technologies (FreeSync and G-Sync) are orthogonal concepts. The former will sync the content to the display, while the latter will sync the display to the content. However, wouldn't it be possible to use both of these features in tandem to get the best of both worlds? This kind v-sync (let's just call it "`display-vrr`" for the sake of semantics) would require mpv to make the assumption that your display's refresh rate is 1:1 identical to the content, which means that the video speed would not be affected.

**Example:** 
Let's say you're watching something that is mastered in 23.976 fps. With `display-vrr`, mpv's display-sync would make the assumption that your display's refresh rate is exactly 23.976 Hz, and FreeSync could sync to that refresh rate without needing any further input from mpv **(assuming your display and/or driver is working as intended)**. If the display doesn't have low frame rate compensation (LFC), or has a bad application of LFC that leads to flickering, then `display-vrr` could be tuned to the nearest integer value that falls between your display's acceptable VRR range; such as 47.952 Hz, 71.928 Hz... etc. 

Intuitively speaking, this behavior should be identical to watching 23.976 fps content on a 23.976 Hz display with `video-sync=display-resample` enabled.

### Advantages this would have over `video-sync=display-resample`
`display-resample` currently works as intended without any issues, and it is a great feature that mpv offers if you want perfect framepacing while watching videos. However, it comes with some drawbacks that are potentially mitigated with `display-vrr`:

1. While being practically imperceptible, `display-resample` leads to changes in video playback speeds and audio pitch, which could be considered undesirable. 
2. With high refresh rate displays, it leads to a noticeable performance impact and power consumption increase.
3. To mitigate 1 and 2, the user has to add multiple custom display modes through their GPU driver for different content types, and switch between them manually or with external scripts. Needless to say, this is purely an issue of convenience and ease-of-use.

`display-vrr` would _theoretically_ be a universal solution to all of these problems, while still taking advantage of `display-resample`'s benefits such as avoiding repeated and dropped frames, and the various .mkv timestamp corrections.

### Advantages this would have over `video-sync=audio` + VRR
I'm speaking completely anecdotally here, but using VRR by itself in mpv doesn't feel good on my setup. It doesn't avoid the issue of repeat frames, and it actually exasperates the problem because repeat frames will stay on screen for a longer amount of time since the display's refresh rate is lower. `display-vrr` should _theoretically_ mitigate this, if my assumptions about how `display-resample` works are correct.

### Addendum
I also wanted to bring up the fact that `display-resample` is based on [RetroArch's display sync](https://github.com/libretro/docs/blob/master/archive/ratecontrol.pdf), and RetroArch offers an option called "Sync To Exact Content Frame Rate" for VRR displays that requires RetroArch's Dynamic Rate Control to be enabled for the best results.

https://www.libretro.com/index.php/upcoming-retroarch-1-7-4-sync-to-exact-content-frame-rate-ideal-for-g-syncfreesync-users/
> This option is the equivalent of forcing x1 speed while still allowing fast forward.
There will be no deviation from the core requested refresh rate alongside sound Dynamic Rate Control (the default RetroArch behaviour outside of this setting).
It should adapt to any refresh supported by your screen and provide smooth scrolling.

I don't understand the technical side of this, but this sounds like `display-resample` with `video-sync-max-video-change=0` where the refresh rate is determined by the content. I don't know if this is related at all to my post, and I understand that video game emulation and video playback are two completely different things, but I wanted to post it regardless since it seems relevant to the discussion.

This post is already getting overly long, but I wanted to cover as much ground as possible since I'm genuinely curious as to why `display-resample` and VRR have to work in opposition to each other.

### Log file
N/A

**Comment by haasn** [comment]
Correct me if I'm wrong, but my understanding is that `display-vrr` would be basically identical to `audio`, except that it will implicitly double frames to meet a `minimum FPS` target, as is often required for VRR displays?

**Comment by philipl** [comment]
Have you tried using `--vf=fps=fps="2*source_fps"`? This will duplicate frames to increase the framerate. In trivial testing for me, it is compatible with `--video-sync=display-resample`, and also works with hardware decoding.

**Comment by Jules-A** [comment]
> Have you tried using `--vf=fps=fps="2*source_fps"`? This will duplicate frames to increase the framerate. In trivial testing for me, it is compatible with `--video-sync=display-resample`, and also works with hardware decoding.

But what if you want interpolation?

**Comment by dokur0chan** [comment]
> Correct me if I'm wrong, but my understanding is that `display-vrr` would be basically identical to `audio`, except that it will explicitly double/triple frames to meet a `minimum FPS` target, as is often required for VRR displays?

Not exactly. I think there's already a feature request related to this (#6137), but I personally don't experience this problem because my display is capable of LFC, meaning that my driver will automatically double and/or triple the refresh rate to meet my VRR range. However, the point of this hypothetical "`display-vrr`" would be to move away from `audio`'s implementation of compensating for A/V clock drifts.

From my understanding of how `display-resample` works, it will adjust the video speed to perfectly fit the display's refresh rate, and then it will let the video occupy every vblank (which is part of the reason why it's such a hefty performance penalty for high refresh rate displays). `display-resample` will then compensate for A/V clock drifts by resampling the audio by inaudible amounts, as described by the Dynamic Rate Control paper that libretro links to. `display-vrr` would basically use this audio-based dynamic rate control to avoid drawing repeat frames, while making the assumption that your display is 1:1 congruent with the content (basically, no video speed adjustments). This assumption would be met by your display's VRR, which is why I said this should _theoretically_ be identical to playing 23.976 fps content on a true 23.976 Hz display while using `display-resample`. It would allow the user to utilize VRR to match the content's refresh rate, without having to manually adjust their monitor's refresh rate.

Right now with `display-resample`'s current implementation, it will always run the video to the maximum refresh rate reported by your display, which is part of the reason why it's orthogonal with VRR. However, if I'm reading RetroArch's description of "Sync To Exact Content Frame Rate" correctly, it utilizes its audio-based dynamic rate control in tandem with VRR by forcing the emulation speed to x1 (`video-sync-max-video-change=0`) and letting the content report the refresh rate, which is (usually) lower than the user's max refresh rate, triggering VRR. 

> Have you tried using `--vf=fps=fps="2*source_fps"`? This will duplicate frames to increase the framerate. In trivial testing for me, it is compatible with `--video-sync=display-resample`, and also works with hardware decoding.

I had never tried combining the FFmpeg frame dupe filter with `display-resample` before, but in my testing, this doesn't achieve what I'm describing since it still runs the video to my max refresh rate (which is 144hz). Using `--vf=fps=` in tandem with `display-resample` only decreases my V-Sync ratio and doesn't affect VRR.

**Comment by philipl** [comment]
If your goal is to avoid running below your monitor's VRR window or avoid bad LFC, then `--vf=fps` will do the trick. (ie: you just want working VRR). So in the 71.928Hz example, it would work, if 71.928Hz is within your display's VRR window (and surely it would be). VRR doesn't care about integers or exact values - it just displays at whatever rate the frames are produced. So as long as you don't want the visual effect of `display-resample` (the actual interpolated merged frames), then `--vf=fps` will do what you want.

If you want the interpolated frames from `display-resample`, and want to hit a specific frame rate that isn't the display frame rate, you are out of luck, but this is, IMHO, a weird desire. The interpolated frames are nothing special and generated using very basic frame merging techniques, so getting a consistent frame rate with `--vf=fps` seems more desirable.

**Comment by Dudemanguy** [comment]
>This assumption would be met by your display's VRR, which is why I said this should theoretically be identical to playing 23.976 fps content on a true 23.976 Hz display while using display-resample. It would allow the user to utilize VRR to match the content's refresh rate, without having to manually adjust their monitor's refresh rate.

I think this is a reasonable idea and perfectly doable. There's technically several video sync modes in mpv, but essentially they fall into two categories: the default one (which is what every player does) and the various display-* ones. Neither of them take into account VRR displays (not really surprising given that they predate before that was even a thing). So the assumptions both modes make aren't really valid for VRR displays. There's a reasonable need for a third category.

With a VRR display within the supported range, we are allowed assume a perfect integer vsync ratio. One of the biggest headaches with video playback (the video's frame rate not matching the refresh rate) is completely solved. Of course, this assumes working drivers and the manufacturer not lying to you but still. With such a display, the ideal is for mpv to simply repaint itself at a constant rate and let the the display/drivers handle it. We shouldn't be touching the frames *at all*. So all of the fancy vsync timing stuff that the `display-*` modes do is completely unnecessary. The framedrop check that the default audo mode does also shouldn't be needed.

Of course, we have to compensate for A/V desync somehow, so the way to handle this is completely on the audio side. i.e. by resampling like we do for the `display-*` modes (or possibly using audio filters like `display-tempo` does). So essentially this new mode becomes a mix of both things. It takes the resampling from `display-resample` but preserves the simple timer-based approached the default audio mode does. A mode like this should have still have perfect video playback with no dropped frames like `display-resample` does. I think this has quite a lot of utility because there are people out there with very high refresh rate displays and running `display-resample` on them is just plain wasteful and taxing on the GPU. The default audio mode is not quite ideal either because it will do A/V desync compensation by touching the video side which is something we shouldn't need to do for on a system with a properly working VRR setup.

I don't think the code changes are actually too bad for this, but unfortunately the VRR on my display is a sick joke so I don't really trust any of my own testing. Well anyways, I'll try to come up with something in the near future that hopefully behaves like how I think it should work.

**Comment by dokur0chan** [comment]
> I think this is a reasonable idea and perfectly doable.

If such a display mode is possible, then it would certainly be a novel idea for a media player at least. 

But yes, maybe my posts have been a little too long-winded, and I might've been mixing up technical jargon that I don't 100% understand, but what you're describing is essentially what I'm talking about. A mode where the video is unaffected, but A/V sync is handled through audio resampling (and hopefully .mkv timestamp jitter is accounted for in this mode as well).

Admittedly, I can already think of one limitation this mode might have, and it's watching videos in windowed mode, since (on my setup) FreeSync doesn't trigger unless I fullscreen the application. However, I don't think this is really an issue because most users probably aren't too hung up about things such as perfect framepacing if they're watching stuff in windowed mode, and if they are, `--video-sync=display-resample` already fulfills their needs.

**Comment by Jules-A** [comment]
> If your goal is to avoid running below your monitor's VRR window or avoid bad LFC, then `--vf=fps` will do the trick. (ie: you just want working VRR). So in the 71.928Hz example, it would work, if 71.928Hz is within your display's VRR window (and surely it would be). VRR doesn't care about integers or exact values - it just displays at whatever rate the frames are produced. So as long as you don't want the visual effect of `display-resample` (the actual interpolated merged frames), then `--vf=fps` will do what you want.
> 
> If you want the interpolated frames from `display-resample`, and want to hit a specific frame rate that isn't the display frame rate, you are out of luck, but this is, IMHO, a weird desire. The interpolated frames are nothing special and generated using very basic frame merging techniques, so getting a consistent frame rate with `--vf=fps` seems more desirable.

Hmm... yeah, it isn't great but nothing else in MPV seemed to work... I haven't actually tried with --vf=fps yet but my use-case is anime where stuff most likely isn't even 23.976 fps so it looks pretty bad with nothing.

**Comment by Dudemanguy** [comment]
@dokur0chan:

We had a bit of a discussion about this is in IRC and perhaps I've gotten a fundamental assumption incorrect, so maybe if you could clarify a bit about the current behavior with `video-sync=audio` that would be great. Just some questions.

1. When you say you experience repeat frames could you elaborate a little more on this? Like, how often does it occur, does it only happen with "heavier" rendering settings, etc.?
2. Do ever experience framedrops? If yes, do you know about how often? Is it only with "heavier" rendering settings?
3. Does using the `--framedrop=no` option change anything at all?

**Comment by dokur0chan** [comment]
@Dudemanguy 
Sure! I don't mind clarifying my original statement, because like I said, it's purely anecdotal and I don't really know how to properly convey it. Part of the reason why I opened the issue was to have a better understanding of these things anyways, I wasn't expecting it turn into a genuine feature request.

>When you say you experience repeat frames could you elaborate a little more on this? Like, how often does it occur, does it only happen with "heavier" rendering settings, etc.?

The "repeat" frames appear like a slight microstutter on my screen, I don't know if it's appropriate to call them judders (and theoretically, VRR shouldn't have any issues with pull-down judder anyways). I can't exactly quantify the amount of times I notice the stutters, because these kinds of things are naturally more noticeable during panning shots rather than still shots. But if we consider the average length of a TV episode to be 20-25 minutes, I might notice it 0-2 times (however, if I were to guesstimate a median, I more often than not never notice it, but when it does occur, it is kinda annoying). It's not baked into the encode, because usually if I rewind and replay the scene, I won't observe the same uneven framepacing during a camera pan. The settings I have don't seem to have any impact on this, as I have a relatively good desktop and I don't often use external glsl-shaders. I typically use everything that comes with `--profile=gpu-hq`.

> Do ever experience framedrops? If yes, do you know about how often? Is it only with "heavier" rendering settings?

No. I don't experience framedrops, and during these "repeat" frames the stats OSD doesn't report a framedrop.

>Does using the --framedrop=no option change anything at all?

Since I don't experience framedrops, I would assume it wouldn't? However, if you think that it might have an impact on my VRR display, I'm willing to daily drive it for a bit and see if I notice any issues.

**Comment by Dudemanguy** [comment]
If you don't drop frames then yeah `--framedrop=no` shouldn't matter.

>The "repeat" frames appear like a slight microstutter on my screen, I don't know if it's appropriate to call them judders (and theoretically, VRR shouldn't have any issues with pull-down judder anyways).

I suppose the key is to figure out what exactly this is and what causes it (mpv or something else). And yes, I wouldn't expect this to be a performance issue if you're merely using `--profile=gpu-hq` and not on a potato.

**Comment by dokur0chan** [comment]
> I suppose the key is to figure out what exactly this is and what causes it (mpv or something else).

Right. I always assumed this was an issue with `video-sync=audio` because it makes no assumptions about the user's display, and on a typical 144 Hz panel playing a 23.976fps video, I'm pretty sure there's some way to calculate the frequency of repeat frames (assuming traditional V-Sync and no VRR). If I remember correctly, it was something like one repeat frame every 40-or-so seconds, but I don't remember the exact details anymore. I thought whatever I was observing was related to that, or some kind of compounding timestamp jitter issue resulting in uneven framepacing, which only gets exasperated by VRR.

Technically, mpv is blacklisted on my drivers (this was actually a fairly recent blacklist on official AMD drivers), but I can force VRR on applications regardless of the blacklist. I can confirm that I can get VRR to kick in with VLC. This is the first time I've used VLC, so I'm not super familiar with it, but I assume by default its V-Sync behaves exactly like `video-sync=audio`. I'll play around with it and report if I notice any issues, but since the "stuttering" happens infrequently, it might take me a while to actually notice anything.

But for the time being, the only other application I can share my anecdotes with is RetroArch, which is part of the reason I brought it up in my addendum (and it's the only application I know of that actually utilizes `display-resample` other than mpv). VRR works quite smoothly with it, or at the very least, I've never noticed these "repeat" frames while playing something like a side-scroller where the screen is scrolling at a constant rate. But again, I understand that video presentation ≠ emulation.

**Comment by llyyr** [comment]
> The "repeat" frames appear like a slight microstutter on my screen, I don't know if it's appropriate to call them judders (and theoretically, VRR shouldn't have any issues with pull-down judder anyways). I can't exactly quantify the amount of times I notice the stutters, because these kinds of things are naturally more noticeable during panning shots rather than still shots. But if we consider the average length of a TV episode to be 20-25 minutes, I might notice it 0-2 times (however, if I were to guesstimate a median, I more often than not never notice it, but when it does occur, it is kinda annoying). It's not baked into the encode, because usually if I rewind and replay the scene, I won't observe the same uneven framepacing during a camera pan. The settings I have don't seem to have any impact on this, as I have a relatively good desktop and I don't often use external glsl-shaders. I typically use everything that comes with --profile=gpu-hq.

Your monitor should have an in-built refresh rate overlay, can you turn it on and see what what refresh rate your monitor reports when you're watching stuff on mpv full screened + VRR with video-sync=audio? Try content of various fps (there are judder tests you can find on the internet at various fps), like 24p, 30 fps, 60fps etc.

**Comment by dokur0chan** [comment]
Yes, I'm aware that most video content is below my VRR range. Usually when I watch 24p content my refreshrate is tripled to 72 Hz, 30 fps is doubled to 60 Hz, etc etc. (I can confirm these from the monitor's OSD). I also have a monitor that has a VRR range of 48 Hz to 144 Hz, coincidentally enough.

> Try content of various fps (there are judder tests you can find on the internet at various fps), like 24p, 30 fps, 60fps etc.

Could you provide a link for some? I've had no luck finding good arbitrary test videos for this. And whenever I come across a random forum post relating to judder, the suggestion is always something along the lines of "just buy this movie, and use the scene at [insert timestamp here] to test for judder", which is just... lol. But if it works, it works I guess.

> This sounds like your monitor doesn't support going down as low as 24 Hz, and your gpu drivers are fucking up multiplying those frames for the VRR range (they do for mine too).

Right. But I don't know if I would classify it as "fucking up", I'm pretty sure this is an intended feature for LFC-capable displays, since not every VRR display is capable of LFC. And from my understanding, if tripling the refresh rate causes judders, it would be noticeable as a constant pattern, and you would be able to measure out the time intervals for said judder, no? I've personally never been able to do this, and it feels more-or-less random for me.

Anyways, through my completely unscientific testing so far, I haven't been able to replicate my stutter/judder/etc. issue with VLC + VRR _yet_. But that doesn't really mean much. I'll keep using VLC some more until I actually observe the problem.

And as always, if anyone else wants to share their experiences with VRR displays and media players, it would definitely help. I would be happy with someone commenting "It Just Werks™" too.

**Comment by Hrxn** [comment]
LFC is what? Low Framerate Content?

**Comment by dokur0chan** [comment]
> LFC is what? Low Framerate Content?

Low frame rate compensation.

**Comment by dokur0chan** [comment]
@llyyr 

Sorry, I have to retract my previous reply's statement, if you could provide a link to some good test videos I would be appreciative. The testing I've done so far has left me extremely confused, and .webm probably isn't the best video container format to test this on... I can perceive something that resembles judder, but it still happens when the video is within my VRR range (60 fps), and the problem seems _significantly worse_ on VLC somehow. If the problem is _this_ bad I feel like I would've noticed it during my random testing so far with VLC+VRR.

**Comment by llyyr** [comment]
Something like [this](https://www.youtube.com/playlist?list=PL1Y97lUkZ4QNs4TvqU5Jw-JalcL1QWUbZ) should be fine. Set your monitor to 60Hz with no VRR and play the 60fps one as a baseline for what to expect, then try it with other refresh rates + VRR. Also include VLC in your testing if you like. For me there's video-sync=audio + VRR for 60fps content is identical to video-sync=audio + no VRR for 60fps content, but that's not the case for anything that's outside my VRR range and the gpu has to multiply frames.

> it would be noticeable as a constant pattern

I'd expect that it happens constantly but is only perceptible on panning/zooming shots, which aren't constant. You could also A-B loop a panning shot from some video instead of what I linked.

**Comment by dokur0chan** [comment]
These are the test videos I used. Specifically, the 23.976 fps, 29.97 fps and 60 fps ones since these framerates map onto real content.

I need to repeat the test patterns a bunch of times, which makes reproducing this fairly tedious, but the juddering does happen. It's hard to tell on the 60 fps one, so I can't really make any definitive judgements on that one. 

However, one thing I do not understand is why the problem is so much worse on VLC. If I have to repeat the test pattern a dozen times on mpv to see it happening once, I can see it happening on VLC _immediately_ and the bad framepacing is constant. The framepacing is horrible even if I disable VRR, drop my monitor down to 60 Hz, and watch the 60 fps test video. However, this isn't a VLC issue tracker, so understandably I don't expect anyone to investigate this.

> I'd expect that it happens constantly but is only perceptible on panning/zooming shots, which aren't constant. You could also A-B loop a panning shot from some video instead of what I linked.

I might do this once I've rested a bit and recovered from the eye strain. Part of the reason why I found the VLC results so shocking is because what I saw doesn't map onto panning/zooming shots in real content, and I never experienced anything that severe.

Unfortunately, perhaps using VRR for media players was never meant to be, and it's only useful for framepacing in video games. However, I still invite other users to share their experiences with VRR.

**Comment by haasn** [comment]
https://github.com/mpv-player/random-stuff/tree/master/vsync is a good source of test clips

**Comment by haasn** [comment]
> Not exactly. I think there's already a feature request related to this (https://github.com/mpv-player/mpv/issues/6137), but I personally don't experience this problem because my display is capable of LFC, meaning that my driver will automatically double and/or triple the refresh rate to meet my VRR range. However, the point of this hypothetical "display-vrr" would be to move away from audio's implementation of compensating for A/V clock drifts.

There should *be* no A/V clock drift with `video-sync=audio` on a VRR display, since the frames can be presented exactly when needed (in theory). hence my confusion about what a `display-vrr` mode would even *do*, to distinguish it from `video-sync=audio`. The only thing I can think of would be software LFC, which may solve some issues if the LFC implementation in the GPU/display are not consistent / result in stutters.

Though to be honest this all confuses me. In theory I have a VRR capable display, so I should do some testing here also. (It just sucks that my main platform, X.org, doesn't support proper VRR and probably never will)

**Comment by philipl** [comment]
VRR works on X just fine with nvidia and mesa if you have one monitor and are not using GNOME (because of a regression in GNOME 44, which they don't see any urgency in fixing)

**Comment by dokur0chan** [comment]
@haasn 
Thank you for the test clips! I didn't realize these were tucked away in the mpv org. 

Times like this make me wish I had a high-speed camera though, or at least a camera that is capable of true slow motion, because I'm just eyeballing it right now. I downloaded the 23.976, 29.97, 30 and 59.94 fps ones for my testing, and I queued all four of them back-to-back as playlist in mpv. It's _really_ hard to notice anything, and I saw maybe like two or three framepacing oddities that I might consider a jitter/stutter/judder/etc. and not just my monitor's low response speed playing tricks on my eyes. But again, each of these test videos are 5 minutes long, and I've basically been looking at my monitor for 40 minutes straight because I repeated the test a second time. I'll have to be locked up in an insane asylum after this...

I also tested the videos out on VLC with VRR, and the bad framepacing is immediately noticeable as with the previous test. I don't know why this happens, just reporting what I see. There must be some internal difference that makes VRR more favorable towards mpv rather than VLC, for some reason.

> There should be no A/V clock drift with video-sync=audio on a VRR display, since the frames can be presented exactly when needed (in theory).

Interesting. This doesn't make intuitive sense to me, but obviously you and the rest of the mpv contributors know more about this subject than I do. Maybe VRR is just fundamentally incompatible with video playback. 

@Jules-A 
> Many monitors, especially non-144hz ones don't actually have LFC, like my Samsung 75hz one that is just 48-75. My use-case includes interpolation as well so I would hope an option like this would allow playback with interpolation at a perfect 3x without having to set a custom resolution and override MPV's framerate estimation (though that seems more like an MPV bug).

I can't speak for interpolation, as its not a feature I personally use. However, since the frames are being blended, you shouldn't be perceiving any juddering to begin with, which means you don't need to set your monitor to an integer value of the content frame rate. `display-resample` and `interpolation` should be working as intended. Also, mpv isn't VRR-aware (I don't think any software is VRR-aware beyond your monitor's OSD), so it shouldn't be able to estimate your VRR refresh rate, as there isn't anything that actually reports it.

**Comment by Jules-A** [comment]
> I can't speak for interpolation, as its not a feature I personally use. However, since the frames are being blended, you shouldn't be perceiving any juddering to begin with, which means you don't need to set your monitor to an integer value of the content frame rate. `display-resample` and `interpolation` should be working as intended. Also, mpv isn't VRR-aware (I don't think any software is VRR-aware beyond your monitor's OSD), so it shouldn't be able to estimate your VRR refresh rate, as there isn't anything that actually reports it. It's not really a bug or an issue.

Maybe it shouldn't but it does... I'm really not sure what it's trying to do but enabling/disabling VRR doesn't do anything to help it, only setting a perfect multiple does.
The bug I was referring to wasn't anything to do with VRR, it's with my custom display resolution. When I set it to 71.928, Windows and MPV dumbly detect it as 71.000 so I have to set override-display-fps to the correct value otherwise I get stutter. 

<details>
  <summary>I'm doing this to set the overrides:</summary>

 ```
[displayfps23.976]
profile-cond=container_fps < 24
profile-cond=container_fps > 23
override-display-fps=71.928

[displayfps24]
profile-cond=container_fps < 25
profile-cond=container_fps >= 24
override-display-fps=72

[displayfps30+]
profile-cond=container_fps > 29
override-display-fps=74.970
```

</details>

but I have to change my resolution manually.

**Comment by dechamps** [comment]
I just stumbled upon this discussion and am following with great interest, as VRR is, at least on paper, the perfect approach to get video at any frame rate (even variable!) to play with perfect frame pacing with no alterations to video nor audio whatsoever.

As far as I know, no PC video player supports VRR playback. Or at least not officially - if it works it's only by accident. It's worth wondering why that is. As a useful data point for discussion, I'd like to bring up [madshi's opinion on this topic from the perspective of madVR](http://forum.doom9.org/showpost.php?p=1271416&postcount=2):

> For perfect motion smoothness, we want the display to show each video frame for *EXACTLY* the right amount of time, which is usually 1000 / 24.000 * 1.001 = 41.708333333333333333333333333333 milliseconds.
>
> FreeSync/G-SYNC would help with video rendering only if they had an API which allowed madVR to specify which video frame should be displayed for how long. But this is not what FreeSync/G-SYNC were made for, so such an API probably doesn't exist (I'm not 100% sure about that, though). Video renderers do not want a rendered frame to be displayed immediately. Instead they want the frames to be displayed at a specific point in time in the future, which is the opposite of what FreeSync/G-SYNC were made for.

The above was [written in 2016](http://forum.doom9.org/showthread.php?p=1787603&highlight=g-sync#post1787603). I am not aware of the situation having changed since then, but I'd certainly love to be proven wrong.

Given the above, I'm a bit surprised to see people discussing VRR in mpv as if it's supposed to work. Could that be because, by sheer luck, mpv happens to present frames with low enough jitter that the above problem is not noticeable in practice? Or does the above not apply to mpv video presentation code paths somehow?

Also, could it be that the reason why it looks so much worse in VLC is because, by sheer misfortune, VLC video presentation calls just happen to have more jitter than mpv's?

**Comment by Jules-A** [comment]
> Given the above, I'm a bit surprised to see people discussing VRR in mpv as if it's supposed to work. Could that be because, by sheer luck, mpv happens to present frames with low enough jitter that the above problem is not noticeable in practice? Or does the above not apply to mpv video presentation code paths somehow?

Jitter is not the problem... This is running at my default 75hz with Freesync and Enhanced sync forced in driver: https://i.imgur.com/bcFOjkH.png Vsync jitter is basically a constant 0.003 but audio is not fully synced and I can notice some stuttering which is why I have to set my resolution manually to 3x... https://i.imgur.com/DlaX7zZ.png

With display-resample of course Freesync won't kick in because it's designed to sync to the monitors max fps, which is why this alternate display-resample mode is being requested...

**Comment by dokur0chan** [comment]
@dechamps 

I am already aware of madshi's previous statements on this topic, and it certainly adds to my skepticism about whether or not VRR is viable for video playback. He does say in the same post that he doesn't own a VRR display to actually test these things, so perhaps he has changed his mind on the topic. The latest post of his I can find about this topic is [from 2018...](https://forum.doom9.org/showthread.php?p=1840046&highlight=freesync#post1840046)

> Great! So we will probably be able to drive these TVs with 120hz *fixed*sync. Looking forward to that.

...where I think the _obvious_ implication is that he was still against VRR at this time. It's a little hard to pick madshi's brain about this topic nowadays since madVR's development has been slowed down, and because madshi doesn't post on doom9 anymore. But I think it's safe to say that he doesn't believe in the viability of VRR for media players.

> Given the above, I'm a bit surprised to see people discussing VRR in mpv as if it's supposed to work. Could that be because, by sheer luck, mpv happens to present frames with low enough jitter that the above problem is not noticeable in practice? Or does the above not apply to mpv video presentation code paths somehow?

My understanding of this topic is very lay, but this does sound like a plausible explanation. Or at the very least, it's the only explanation for why this discrepancy between mpv and VLC exists so far.

**Comment by dechamps** [comment]
For what it's worth, I personally suspect that madshi is overly pessimistic on this topic. My understanding is that he didn't even try to at least prototype VRR support in madVR, probably because it was never a topic of much interest to him.

To be clear, I do agree with madshi that the way VRR works with frame presentation APIs means that there is probably no way to make it work *perfectly*, but it seems plausible that it can be made to work *well enough*. For example by presenting the frame in a dedicated thread running at the highest possible priority and using tricks (e.g. scheduling the thread in advance and then busy-looping, I don't know) to keep jitter as low as possible (<1 ms worst case is probably good enough). This would arguably be a giant hack (which might be why madshi didn't want to do it), but it might actually work in practice. I'd be excited for it to be prototyped in mpv somehow. Perhaps we're already there or almost there, given the discussion taking place in this thread and the seemingly widespread belief that mpv already works with VRR?

I'd certainly love for someone intimately familiar with mpv frame presentation/VSync code paths (I certainly am not) to chime in.

**Comment by haasn** [comment]
Why would vsync jitter be a problem with VRR? If vsync jitter caused issues with VRR, that would defeat the whole purpose of VRR even existing. Do you know what else has variable frame times? Video games, i.e. *exactly* the use case VRR is designed to handle! The whole point of VRR is to *not* require a constant, fixed framerate.

madshi's statement has to be interpreted in context - madVR uses custom modelines to target an exact pixel clock that is tuned to closely match the content. Of course, this only works on windows, requires fullscreen exclusive mode, and a display that can be timed to such arbitrary modelines to begin with (like.. gosh, a *VRR-capable display*). So obviously madshi doesn't think VRR provides much value. (Although even then, VRR would allow A/V drift compensation without resampling, so...)

For mpv none of the above applies.

> I just stumbled upon this discussion and am following with great interest, as VRR is, at least on paper, the perfect approach to get video at any frame rate (even variable!) to play with perfect frame pacing with no alterations to video nor audio whatsoever, out-of-the-box with zero configuration. As far as I know, no PC video player supports VRR playback. [...]

I'm not aware of any environments in which VRR support requires the video player to do something *special* to enable it. Although on Linux, you might be running into [the mesa VRR blacklist](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/20701?commit_id=fed462a011d74d1be569d3b6cd3efcd75ac2d54b), which should be removed on recent enough versions.

So it should just work out of the box. If it doesn't, I don't know - blame your drivers of environment, but not the design of VRR, which doesn't require anything special for video players to take advantage of.

If I had to guess, I'd say that one possible defect is because of bad LFC algorithms, hence my suggestion to make `display-vrr` just ensure that we're duplicating/dropping frames as needed to match the minimum/maximum fps targets.

**Comment by haasn** [comment]
So, after some extensive testing involving a [rewrite of my plplay test app](https://code.videolan.org/videolan/libplacebo/-/merge_requests/503) and a [slow motion video camera](https://en.wikipedia.org/wiki/IPhone), I've come to the following conclusions:

1. VRR works "out of the box" on Wayland + KDE Plasma, after enabling it in the display options.
2. The app doesn't need to do anything special to take advantage of it, it's enabled automatically (in both fullscreen *and* windowed!)
4. **Moving the cursor** sometimes introduces a slight hiccup as the cursor update pushes the display from content-synced mode into maximum refresh, but this is not hugely noticeable at 175 Hz.
5. Testing with a [specialized display test app](https://github.com/Nixola/VRRTest) confirms my display is not skipping or repeating frames, even as the framerate varies.
6. LFC on my display+GPU combo works fine for normal content, but *may* stutter/flicker sometimes when the framerate suddenly changes (e.g. VFR going from 24 Hz to 30 Hz). This effect seems transient though. I haven't yet tested it on "true" VFR content (where every frame is timed differently, rather than being merely a glued-together series of CFR segments)
7. ~30 Hz content sometimes likes bouncing between 60 Hz (frame doubling) and 90 Hz (frame tripling), but this doesn't appear to cause any visible issues.
8. `display-resample` continues working perfectly on VRR modes, it will simply sync to the maximum rate supported by the monitor, exactly as is the case in traditional CFR mode.

So in conclusion I still don't understand what this issue is about. It all just works(tm) on my end, with the possible exception of LFC, but I suspect that monitor implementation varies wildly here.

Can people with issues confirm that everything works fine as long as the display is within the expected framerate range? (See also the above VRRTest app, which is hugely helpful in testing)

**Comment by haasn** [comment]
Okay... time for a stupid question:

**Are you launching mpv from the command line or the graphical environment?**

By default, mpv prints a terminal status message about the current playback position, and also any logs (as needed). On my end (kde plasma wayland), this causes a terminal repaint **even if the mpv window is fullscreened in front of it**!!

Solution? Adding `--really-quiet` makes playback in `mpv` as buttery smooth as in `plplay` for me. Without this option, I also get jittery fps and stutter all over the place, when launching mpv from the terminal (but not from the `.desktop` file)....

Conclusion: Compositor bug, at least in plasma wayland.. the terminal being hidden behind the mpv window should NOT mess up all frame timings. Actually, the terminal updating should not mess up frame timings ever, IMO, but YMMV.

**Comment by dokur0chan** [comment]
Right. I've always found madshi's hard-line statement on this to be pretty weird, as I don't understand why VRR wouldn't just werk™ with video applications beyond just driver blacklists.

The application you linked is awesome! I've actually never seen an application trigger FreeSync in windowed mode before (at least, not on Windows). I've done some testing for about an hour now, and I've even attempted to use the slow motion mode on my phone camera (it kinda sucks). I haven't been able to see any issues both when the refresh rate is in my VRR range, and when it's outside of my VRR range. I also haven't been able to see any visible issues when I do the 24 Hz - 30 Hz fluctuation test to simulate something like VFR content. By looking at my OSD, it seems like the driver likes to flip-flop a lot on deciding whether it wants to double, triple, or even quadruple the refresh rate when doing some fluctuation tests, but this doesn't seem to cause any visual issues. So seemingly, everything just werks™

> Are you launching mpv from the command line or the graphical environment?

I'm on winblows. I usually don't launch mpv from a terminal unless I'm playing around with yt-dlp, and I didn't do so for any of the video test files. But good to know.

In conclusion, assuming everything is working and we've confirmed that mpv just works OOTB, then I can't really diagnose what my issue with VRR is on mpv. What I'm witnessing during video playback could just be random driver hiccups, random DE or DWM hiccups, general encode timestamp randomness, video-induced psychosis... god knows what at this point. But at the very least, this discussion has helped clarify my initial question on whether or not what I was seeing was A/V sync related.

**Comment by Jules-A** [comment]
Also on Windows, that app is weird, in Fullscreen vsync needs to be enabled for Freesync to work and adding profile in driver to specifically enable Freesync doesn't help. In windowed, Freesync is activated with Vsync enabled or disabled and even when I specifically disable it for the app in driver :/

**Comment by Dudemanguy** [comment]
I'm not sure if there's any other unlucky soul out there like me, but thanks to haasn's tip, I've discovered that VRR actually works for me. But only on plasma wayland and with that argument (never thought to use it before so naturally I just thought it was broken). It doesn't work in sway (despite adaptive sync being on) nor at all in xorg (even though I only use one monitor and I have adaptive sync enabled). I don't know why this is the case for me since I know there are people that use VRR in sway fine.

**Comment by Hrxn** [comment]
> Solution? Adding `--really-quiet` makes playback in `mpv` as buttery smooth as in `plplay` for me. Without this option, I also get jittery fps and stutter all over the place, when launching mpv from the terminal (but not from the `.desktop` file)....
> 

I wonder if it's possible to reproduce a similar behavior in Windows, terminal repaints etc. are a potential issue here as well, I think?

**Comment by dokur0chan** [comment]
> I wonder if it's possible to reproduce a similar behavior in Windows, terminal repaints etc. are a potential issue here as well, I think?

Can't repro personally, or at the very least, I don't notice anything obvious when using the Win11 terminal.

**Comment by dechamps** [comment]
> Why would vsync jitter be a problem with VRR? If vsync jitter caused issues with VRR, that would defeat the whole purpose of VRR even existing. Do you know what else has variable frame times? Video games, i.e. exactly the use case VRR is designed to handle! The whole point of VRR is to not require a constant, fixed framerate.

I think it's a question of degree. I don't think anyone expects games to be *perfectly* smooth - very few games have perfect frame pacing, and some hiccups here and there are accepted as a fact of life. That's not true for video - a typical hardware video player presents frames at an absolutely perfect cadence and any kind of jitter is considered unacceptable, so that's the benchmark that a PC player will be compared against. In other words: the bar is higher for video than it is for games, IMHO.

> madshi's statement has to be interpreted in context - madVR uses custom modelines to target an exact pixel clock that is tuned to closely match the content. Of course, this only works on windows, requires fullscreen exclusive mode, and a display that can be timed to such arbitrary modelines to begin with (like.. gosh, a VRR-capable display). So obviously madshi doesn't think VRR provides much value. (Although even then, VRR would allow A/V drift compensation without resampling, so...)

I'm not entirely sure what you're referring to. madVR does not use "custom modelines - at least not in the renderer itself. It does come with a separate and independent [utility](http://madvr.com/crt/CustomResTutorial.html) (which was added relatively late in madVR history) for creating custom modes on the GPU driver, but that has to be done separately in a somewhat tedious and time-consuming process - the renderer does not do it on the fly and is not involved in that process at all. Since none of this is done during video playback, it does not "require fullscreen exclusive mode". It's also not specific to Windows - in fact, my brother who runs mpv on Linux manually crafted and added custom modelines to his X server config to achieve the exact same result.

> I'm not aware of any environments in which VRR support requires the video player to do something *special* to enable it.

I never said *enabling* VRR required anything special. I have no problem believing that it *works*. However I would not intuitively expect it to work *well* given that, as far as I understand, any thread scheduling jitter on the thread responsible for frame presentation will directly translate to frame jitter on the screen - something that's not true when using fixed VSync. This thread scheduling jitter cannot be eliminated given neither Windows nor Linux are real-time OSses - they do not make hard guarantees wrt thread scheduling.

If you're saying that this jitter is low enough in practice that it doesn't matter, then I'm glad to hear it. I'd still be worried about worst-case 99th-percentile long tail events though, like a frame being delayed by a few milliseconds in particularly sensitive content such as a wide panning shot due to some freak event like some background system process waking up.

I have a VRR-capable display (LG G1) and some DIY measuring equipment (using a photodiode and an oscilloscope) to measure actual frame presentation intervals directly off the physical light output of the display. I'm not home at the moment, but when I'm back I might do some experiments with mpv in VRR mode to see if I can put some numbers on frame presentation jitter in fixed VSync vs. VRR mode.

**Comment by haasn** [comment]
@dechamps Okay, I understand now what you're worrying about - you want to reduce vsync jitter, whereas my comments / DS are concerned about dropped/duplicated frames.

Yes, it would definitely be better to try and present frames with as low jitter as possible if you want to maximize perceived motion smoothness. Having some numbers here would be a huge benefit. Mpv itself can already measure vsync timing consistency, though not (yet) in audio mode which is where these measurements would be most interesting..

I think we misunderstood each other because I am not even worrying about such relatively minor deviations in frame time when the alternative (VRR off) causes frames to be repeated or missed entirely - a jitter of 100%+. So having VRR massively improves the status quo. When you said "jitter is a problem" I thought you meant "because it will cause dropped frames", not "because it may translate to visual unsmoothness".

But now I understand what you want `display-vrr` to do - essentially, to schedule the frame internally and delay until the scheduled time using a sleep call (instead of blocking on swapbuffers), adjusting the scheduled timestamp slightly forwards or backwards to maintain A-V sync? (Or better yet, sleep based on the audio clock, though that may be harder or impossible to do effectively)

This would probably fit into the DS framework and also eliminate microstutter due to e.g. shader recompilation because we would render the frame before sleeping until the desired presentation time. I don't think a dedicated thread would be necessary, since the VO/DS code already runs on its own thread - just need to make sure to do all CPU work before submitting the frame to the swapchain.

**Comment by dechamps** [comment]
Agreed @haasn - glad we're on the same page.

> I think we misunderstood each other because I am not even worrying about such relatively minor deviations in frame time when the alternative (VRR off) causes frames to be repeated or missed entirely - a jitter of 100%+. So having VRR massively improves the status quo. When you said "jitter is a problem" I thought you meant "because it will cause dropped frames", not "because it may translate to visual unsmoothness".

You are correct - I was implicitly comparing VRR against "perfect" playback where (1) there are no dropped nor repeated frames *and* (2) frames are presented with perfectly constant intervals between them down to the sub-millisecond level (no jitter). The point I was trying to make was, a VRR-based solution will be somewhat "less than perfect" when compared against this ideal, at least on paper, because of the potential frame presentation jitter issues. (My understanding is that this is precisely the reason why madshi/madVR doesn't like VRR.) I do agree, however, that VRR is likely way better than a situation where the display refresh rate doesn't even match a multiple of the content FPS at all. Sorry for the confusion.

> But now I understand what you want `display-vrr` to do - essentially, to schedule the frame internally and delay until the scheduled time using a sleep call (instead of blocking on swapbuffers), adjusting the scheduled timestamp slightly forwards or backwards to maintain A-V sync? (Or better yet, sleep based on the audio clock, though that may be harder or impossible to do effectively)

Perhaps. I'm afraid I'm not familiar with mpv internals and have no idea what its render code paths look like, so I can't comment on your proposed solution. I was just pointing out that whatever solution is used, it would be nice if it could keep actual frame presentation jitter (as measured from display output) as low as possible.

**Comment by haasn** [comment]
@dechamps I looked at the mpv timing code and it turns out that this is *already* what mpv does in audio timed mode:

```c
        stats_time_start(in->stats, "video-draw");

        if (vo->driver->draw_frame) {
            vo->driver->draw_frame(vo, frame);
        } else {
            vo->driver->draw_image(vo, mp_image_new_ref(frame->current));
        }

        stats_time_end(in->stats, "video-draw");

        wait_until(vo, target);

        stats_time_start(in->stats, "video-flip");

        vo->driver->flip_page(vo);
```

So in conclusion, we just need to make sure that `draw_image` only prepares the image (and handles all rendering etc.), while `flip_page` is what actually makes it visible on-screen.

I pushed f9dc695b580c394bf4f9833d36e91b7fcbe009ea which should make `--vo=gpu-next` handle this correctly, so in theory vsync jitter should already be as low as we can realistically make it.

In summary, I'm still not convinced there's anything we specifically need to be doing different.

**Comment by dechamps** [comment]
@haasn Thanks, I'll make sure to give it a try and set up some measurements as soon as I'm able.

As a side note: I got curious and while taking a quick look at libplacebo's high-level interfaces I stumbled upon [`pl_swapchain_latency`](https://code.videolan.org/videolan/libplacebo/-/blob/fcba744b7fd74af98a0455fb1f056fe851cc7cf5/src/include/libplacebo/swapchain.h#L49), which reminds me of "present several frames in advance" options in madVR. Apologies if this is a dumb question, but could this cause problems here? It seems that VRR could get confused if multiple frames are submitted in advance, as it's unclear when the frames are supposed to be sent to the display?

**Comment by haasn** [comment]
@dechamps this is not relevant in audio-timed mode, only for DS (where we drive the render loop as fast as possible, relying on the display to block in order to measure the progress of time). In audio timed mode there should never be multiple frames submitted in advance, precisely because of the fact that we sleep until the appropriate target timestamp before even beginning work on the next frame. (This is not ideal because it may prevent the GPU from starting work on the next frame early if the current frame took less time than expected, but current API limitations, especially on GL, prevent us from doing something much cleverer here - and it doesn't matter in the event that all frames take roughly the same amount of time to render)

Incidentally, I am currently in the process of scrapping the entire swapchain API in libplacebo and redesigning it from scratch, if you want to take a look at that development I should have a draft MR up soon.

**Comment by dechamps** [comment]
Thanks. One caveat is my setup is very much Windows-focused. We might be able to combine efforts to make sure this works well on all platforms. (I don't think I'd have the skills to contribute code, but I should be able to help verify the resulting behaviour.)

While continuing to peruse libplacebo code I also came across the [D3D11 present call](https://code.videolan.org/videolan/libplacebo/-/blob/fcba744b7fd74af98a0455fb1f056fe851cc7cf5/src/d3d11/swapchain.c#L257) (which, contrary to Vulkan, is done at swap time, not submit time), and I noticed that it doesn't seem to follow [Microsoft recommendations for VRR](https://learn.microsoft.com/en-us/windows/win32/direct3ddxgi/variable-refresh-rate-displays#variable-refresh-rate-displaysvsync-off) which suggest `DXGI_PRESENT_ALLOW_TEARING` should be passed. Not sure how big of a problem this is.

> if you want to take a look at that development I should have a draft MR up soon

I think you greatly overestimate my skills when it comes to graphics/GPU programming 😅 Do feel free to CC me on any related proposals though, as I am quite curious in how this works under the hood.

**Comment by haasn** [comment]
> I think you greatly overestimate my skills when it comes to graphics/GPU programming sweat_smile Do feel free to CC me on any related proposals though, as I am quite curious in how this works under the hood.

I threw a draft up [here](https://code.videolan.org/videolan/libplacebo/-/merge_requests/507)

**Comment by dokur0chan** [comment]
Slightly off-topic, but since the differences between D3D11 and Vulkan are being discussed, I wanted to bring up the fact that on my system (AMD GPU + Windows), VRR doesn't seem to work with `--gpu-api=vulkan`, and only works with `--gpu-api=d3d11`. 

It used to work, however, I need to clarify that this is absolutely _not_ an mpv/libplacebo regression, and is some kind of GPU driver related change and/or regression. The behavior is something along these lines:

1. Enable VRR for mpv.exe (only if blacklisted)
2. Launch mpv using the Vulkan backend
3. Select a video file with typical content frame rates (23.976, 24.000 fps etc.)
4. Open monitor OSD, check refresh rate
5. FreeSync is triggered for about 5-10 seconds, but then disables itself, and the monitor reverts back to max refresh rate
6. Restart mpv and repeat
7. Now FreeSync won't even trigger for the first 5-10 seconds, and remains disabled until you clear your driver profile for mpv.exe (but even then, you will simply repeat the procedure from step 5 and onward if you reapply the driver profile)

Again, this is 100% driver related because old versions of AMD drivers didn't used to do this, and using an older build of mpv doesn't revert the issue. I have never bothered to pinpoint what driver version causes this mostly because I stick to `d3d11` anyways. But if anyone comes across this thread and is wondering why VRR doesn't work with Vulkan, then this might help clarify the source of the problem.

**Comment by Jules-A** [comment]
> I threw a draft up [here](https://code.videolan.org/videolan/libplacebo/-/merge_requests/507)

This seems to be different to what the OP was asking for and what I'm also after. While it may solve the OP's issue, it doesn't solve it for people with no LFC or those wanting to use interpolation. Basically I'm just after a mode that will play content as many times as it can fit in to my refresh rate at an exact integer rate that also works with interpolation. vf=fps=fps="2*source_fps" will avoid LFC issues but it won't work with interpolation and what happens if the source vid is 60fps on my 75hz monitor?

**Comment by dokur0chan** [comment]
> This seems to be different to what the OP was asking for and what I'm also after.

Just to be clear, this wasn't _necessarily_ a feature request. As revealed in this discussion, I misdiagnosed my issue. My idea for `display-vrr` would probably not solve it. So I have more-or-less gotten my answer throughout this thread. If this new swapchain API rewrite solves my problem, then that would be awesome! But if it doesn't, then that's fine as well. `display-resample` already works perfectly.

Although, something that would probably be nice to have in mpv is custom modeline support. I know that this is already achievable through [scripting and using stuff like xrandr](https://gitlab.com/lvml/mpv-plugin-xrandr), but there's no clean solution for this in Windows. The only external scripting tool that allows for refresh rate changes is nircmd, and nircmd only supports whole numbers, which kind of defeats the point of avoiding video resampling. nircmd is also kinda bloat if the only feature you utilize with it is scripting your refresh rates. However, it's still the only solution you can use if you have something like an super-duper meme-tier 360 Hz monitor, and you don't want to burn your house down by using `display-resample` and `interpolation`.

> dokur0chan, Vulkan has always had tonnes of issues in MPV, even going back over a year, I don't remember it ever playing well compared to d3d11.

`vulkan` works fine for me in mpv and has significantly faster shader compilation. It has higher V-Sync jitter compared to `d3d11` (0.025~ vs. 0.002), but in practice it doesn't make a difference in video playback. Like I said, the VRR thing is 100% driver related and is no fault of mpv.

**Comment by haasn** [comment]
> It has higher V-Sync jitter compared to d3d11 (0.025~ vs. 0.002)

The swapchain refactor I linked will fix that.

**Comment by CogentRedTester** [comment]
Just thought I'd add my experience trying to get VRR working with mpv. I am on Windows 10 and am using an Nvidia GPU, I've been testing with a (very bad, not even a refresh overlay) freesync monitor and a (slightly better) Samsung VRR TV. I have used the VRR Test app posted above to confirm that VRR is working on the displays, and the TVs refresh counter is accurately reflecting the fps of the app (though it does seem to be doing some automatic frame doubling). I have activated the Nvidia G-Sync compatibility indicator which adds an overlay for VRR windows.

I have also used the Nvidia performance overlay to confirm that when `video-sync=audio` the fps output by mpv is the actual framerate of the video, or whatever `vf=fps=` is set to, while `video-sync=display-resample` shows the maximum refresh of the display.

This is where it gets a little weird. I spent a couple of hours experimenting with various settings and I could not get the G-Sync indicator to activate or the TVs refresh counter to change. However, while typing up this comment I launched mpv again with my standard config and it actually did work. So either I'm a complete idiot, or my display setup is extremely fragile. One thing I did do was that in the Nvidia Control Panel's program-specific settings I re-added mpv (even though it was already in the list) and I toggled the GSync-Compatability. No idea if that did anything.

Anyway, here is what I have found. When using `--no-config`, I can get VRR to activate only when `gpu-api` is set to `vulkan` or `d3d11`. However, my standard config only works with `vulkan`. After some experimentation I have found that `d3d11` stops working with VRR when `hwdec=auto`. Setting `hwdec` to `no` fixed the issue. Weirdly setting `hwdec` to `no` after mpv had started the video did not re-enable VRR, in fact having `hwdec` enabled at any time during playback seemed to break VRR.

As for the quality of the VRR, I'm not sure I trust my eyes enough to make a judgement on how smooth the motion is. I can't even be sure it's better than VLC without VRR. Though as I said my monitor is very bad.

> Basically I'm just after a mode that will play content as many times as it can fit in to my refresh rate at an exact integer rate that also works with interpolation. vf=fps=fps="2*source_fps" will avoid LFC issues but it won't work with interpolation and what happens if the source vid is 60fps on my 75hz monitor?

Not sure about interpolation (and I don't know why you would need it since my understanding is that VRR makes it unnecessary), but I believe you can emulate LFC using conditional profiles. For my 48-75Hz VRR display I have these profiles:

```properties
[vrr-frame-tripling]
profile-cond= user_data.display_name == 'VX2257' and fullscreen and math.floor(container_fps) >= 20 and math.floor(container_fps) < 25
profile-restore=copy-equal
vf-append=fps=source_fps*3
video-sync=audio

[vrr-frame-doubling]
profile-cond= user_data.display_name == 'VX2257' and fullscreen and math.floor(container_fps) >= 25 and math.ceil(container_fps) < 37
profile-restore=copy-equal
vf-append=fps=source_fps*2
video-sync=audio

[vrr-singe-frame]
profile-cond= user_data.display_name == 'VX2257' and fullscreen and math.floor(container_fps) >= 37 and math.ceil(container_fps) < 75
profile-restore=copy-equal
video-sync=audio
```

This doesn't work with VFR content, but it should still be an improvement. The `user_data.display_name` value is set by my [display-name](https://github.com/CogentRedTester/mpv-scripts/blob/master/display-name.lua) script. I set it up since only one of my monitors supports VRR.

**Comment by Jules-A** [comment]
> Not sure about interpolation (and I don't know why you would need it since my understanding is that VRR makes it unnecessary), but I believe you can emulate LFC using conditional profiles. For my 48-75Hz VRR display I have these profiles:

My main usecase is anime where the source vid's real framerate is often lower causing things to get choppy even if MPV is playing the file correctly. While MPV's interpolation isn't great, I find it better than not using any at all.

> This doesn't work with VFR content, but it should still be an improvement.

It would if I wasn't using interpolation.

**Comment by haasn** [comment]
> > Not sure about interpolation (and I don't know why you would need it since my understanding is that VRR makes it unnecessary), but I believe you can emulate LFC using conditional profiles. For my 48-75Hz VRR display I have these profiles:
> 
> My main usecase is anime where the source vid's real framerate is often lower causing things to get choppy even if MPV is playing the file correctly. While MPV's interpolation isn't great, I find it better than not using any at all.
> 
> > This doesn't work with VFR content, but it should still be an improvement.
> 
> It would if I wasn't using interpolation.

Does interpolation at e.g. 175 Hz really make a difference vs interpolation at 168 Hz (clean multiple of 24)?

**Comment by Jules-A** [comment]
> Does interpolation at e.g. 175 Hz really make a difference vs interpolation at 168 Hz (clean multiple of 24)?

No clue, I only have a 75hz monitor and usually play 23.976 content, rarely 24.000 or 29.970. To me it just doesn't seem right when it's not a clean multiple though maybe it makes less of a difference at higher frame-rates but I wouldn't know.

**Comment by dokur0chan** [comment]
Closing since I already got an answer to my question earlier in the thread. Unfortunate that I wasn't able to diagnose my issue, but at least I cleared up some of the misconceptions I might have had on this topic.

**Comment by dechamps** [comment]
> I have a VRR-capable display (LG G1) and some DIY measuring equipment (using a photodiode and an oscilloscope) to measure actual frame presentation intervals directly off the physical light output of the display. I'm not home at the moment, but when I'm back I might do some experiments with mpv in VRR mode to see if I can put some numbers on frame presentation jitter in fixed VSync vs. VRR mode.

> Yes, it would definitely be better to try and present frames with as low jitter as possible if you want to maximize perceived motion smoothness. Having some numbers here would be a huge benefit. Mpv itself can already measure vsync timing consistency, though not (yet) in audio mode which is where these measurements would be most interesting..

Well I thought this would just be a quick "couple of week-ends" kind of project.

Turns out, it wasn't.

On that note, I do not thank you, @haasn.

I do not thank you because, just through this discussion alone, you [nerdsniped](https://xkcd.com/356/) me so hard that I literally ended up spending *6 months* building a full-blown video presentation timing measurement framework. That's, like, not cool, man.

Anyway, jokes aside, let me present **[videojitter](https://github.com/dechamps/videojitter)**, the video cadence measurement toolbox. I hope you'll find it more suitable than an iPhone camera and a bunch of apps 😁

The tl;dr of videojitter is: point a light sensor to a display, play a test video, record the light sensor signal, plot the results. Because this is measuring physical light output, it's impossible for the playback system to "lie". With the right setup, videojitter can measure frame presentation timing with a somewhat ridiculous degree of accuracy (about 18 *micro*seconds standard deviation in the measurements that follow) - much higher than human perception, as far as I know.

Anyway, let's move on to actual measurements, since this is what we were all waiting for.

# Test setup

- Display: LG 77G1 OLED TV, firmware 03.33.11, connected through HDMI 2.1 (display mode: 3840x2160, 119.88 Hz, 10-bit, HDR)
- NVidia RTX 3080 Ti drivers 546.65
- Windows 11 23H2 22631.3007 (fresh install)
- mpv 0.37.0 1c9c2f5 (shinchiro Windows build)
- mpv command line: `--fs --no-config --vo=gpu-next`
- VRR was disabled/enabled by turning G-Sync on/off globally in the NVidia control panel. The effect was confirmed by looking at the LG VRR info menu.

# 60 FPS

## Fixed VSync

![report](https://github.com/mpv-player/mpv/assets/866859/2a438ecf-5b9b-4da4-bf21-52659f2becc7)
[(spec, raw recording, mpv log)](https://github.com/dechamps/videojitter/tree/2c0978999fb1b3f92cee9dacb8ca8772ff8fef3a/videojitter_test/cases/mpv_59p_at_119hz)

## VRR

![report](https://github.com/mpv-player/mpv/assets/866859/90cd11a3-5109-445e-bda4-11f7eb823200)
[(spec, raw recording, mpv log)](https://github.com/dechamps/videojitter/tree/2c0978999fb1b3f92cee9dacb8ca8772ff8fef3a/videojitter_test/cases/mpv_59p_vrr)

# 24 FPS

## Fixed VSync

![report](https://github.com/mpv-player/mpv/assets/866859/57c6c23f-3963-4b98-8692-c4c78d3ddb5d)
[(spec, raw recording, mpv log)](https://github.com/dechamps/videojitter/tree/2c0978999fb1b3f92cee9dacb8ca8772ff8fef3a/videojitter_test/cases/mpv_23p_at_119hz)

## VRR

![report](https://github.com/mpv-player/mpv/assets/866859/849d5f64-1db4-4d2c-90cb-ac6372fc0466)
[(spec, raw recording, mpv log)](https://github.com/dechamps/videojitter/tree/2c0978999fb1b3f92cee9dacb8ca8772ff8fef3a/videojitter_test/cases/mpv_23p_vrr)

# Discussion

The fixed VSync results serve as a baseline and show a perfect reference. Despite mpv running in `--video-sync=audio` mode, I was lucky enough that potential audio/video clock skew did not impact any frames on these measurements. (Surely it would have if I had taken longer measurements.)

The VRR results are exactly what I expected to see, and reveals the phenomenon we predicted through the discussions in this thread: random frame presentation jitter, presumably caused by imperfections in the timing of mpv present calls.

VRR causes frame presentation jitter of about ~800-900 µs standard deviation, which is a somewhat better result than I expected, although I suspect it would vary a lot depending on software/hardware setup.

The worst outlier is located 17 seconds into the 24 FPS measurement, where a frame is displayed for only 36.9 ms instead of the expected 41.7 ms. I'm honestly not sure if that kind of outlier is perceptually relevant - 4.8 ms seems borderline to me. Personally I would prefer not to see >2 ms outliers just to have some safety margin. Maybe software tricks can be implemented in mpv to improve on this result.

I need to add the caveat that these are relatively short 1-minute measurements. It may be worthwhile to do longer measurements (e.g. 1 hour) as, statistically speaking, it will increase the likelihood of a perceptually relevant "freak event" outlier.

**Comment by haasn** [comment]
Wow, super cool and useful measurements! Thanks for investigating this so deeply.

Have you tried putting the vo present thread onto high/realtime scheduling priority?

**Comment by Jules-A** [comment]
Have you tried testing with AMD? What does MPV say in OSD as vsync jitter? I see this with VRR enabled: 
![image](https://github.com/mpv-player/mpv/assets/1760158/389e48f2-3ccc-461f-aba4-105c8b88e117)
but how accurate is MPV's measurements?

**Comment by dechamps** [comment]
> Have you tried testing with AMD?

I don't have an AMD GPU on hand. Feel free to make your own measurements - I took great care to document videojitter in an excrucitating amount of detail and you can [build your own instrument](https://github.com/dechamps/videojitter/blob/master/INSTRUMENT.md) cheaply and easily.

> What does MPV say in OSD as vsync jitter? I see this with VRR enabled and displayresample:

I've never tried the other `--video-sync` options. There must be something wrong somewhere because if I try the setup as described above with VRR on and `--video-sync=display-resample` then the visual result is obviously unusable garbage, mpv complains about A/V desync , and the reported VSync jitter of about ~0.700.

> Have you tried putting the vo present thread onto high/realtime scheduling priority?

Not sure if there's an easy way to do that specifically for the present thread (I don't really have time to dig deep into mpv vo code these days). I just tried switching the entire mpv process to real time priority, and also switched the Windows power plan settings to "High Performance". The resulting measurements are virtually identical to the ones I've just shown.

**Comment by llyyr** [comment]
> VRR enabled and displayresample

You've already been told that display-resample doesn't do whatever you think it does. https://github.com/mpv-player/mpv/issues/6137#issuecomment-1895935179

> I've never tried the other `--video-sync` options. There must be something wrong somewhere because if I try the setup as described above with VRR on and `--video-sync=display-resample` then the visual result is obviously unusable garbage, mpv complains about A/V desync , and the reported VSync jitter of about ~0.700.

Ignore trolling. `display-resample` and VRR are completely incompatible with each other, as stated in the OP of this issue.

**Comment by dechamps** [comment]
> I know I was the one who originally made the claim that VRR has presentation issues on mpv, but a ~5 ms deviation is quite low all things considered.

Keep in mind this is only one measurement from one system (and a short one, at that). I would expect this kind of jitter to be heavily influenced by a variety of hardware and software factors. I suspect there are many reasons why mpv may fail to precisely time a present call, a lot of which are not under mpv's control. If you want to be sure, feel free to make a measurement of your own system and we can compare :) 

I think one way to determine the threshold of just noticeable difference (JND) might be to deliberately induce jitter and then see at which point does it become noticeable on something like a worst-case panning shot, scrolling credits, or similar. (A similar question that I often ponder is: at 120 Hz fixed, if a 24p frame misses a VSync and is displayed one vsync (8 ms) early or late, can anyone actually tell, or is the variation too small to be perceptible? What about 240 Hz?)

**Comment by haasn** [comment]
> > I know I was the one who originally made the claim that VRR has presentation issues on mpv, but a ~5 ms deviation is quite low all things considered.
> 
> Keep in mind this is only one measurement from one system (and a short one, at that). I would expect this kind of jitter to be heavily influenced by a variety of hardware and software factors. I suspect there are many reasons why mpv may fail to precisely time a present call, a lot of which are not under mpv's control. If you want to be sure, feel free to make a measurement of your own system and we can compare :)
> 
> I think one way to determine the threshold of just noticeable difference (JND) might be to deliberately induce jitter and then see at which point does it become noticeable on something like a worst-case panning shot, scrolling credits, or similar. (A similar question that I often ponder is: at 120 Hz fixed, if a 24p frame misses a VSync and is displayed one vsync (8 ms) early or late, can anyone actually tell, or is the variation too small to be perceptible? What about 240 Hz/4 ms?)

I think a good way of testing this would be to produce two 120 Hz video files, one with a perfect 5:1 24 Hz video cadence, and another one with occasional 6:1/4:1 (or 4:1/6:1) frame pairs). Then we could do some home-grown ABX measurements.

**Comment by Jules-A** [comment]
I'm not seeing what you guys are, VRR doesn't seem to cause the same issues for me with display-resample. AFAIK only people with Nvidia cards reported that conflict though I did have issues with it in the past. I know it's not optimal (hence my desire for a solution to this issue) but for whatever reason enabling VRR with display-resample causes the reported monitor refresh-rate to be much more stable, without it, it drifts quite a bit.

**Comment by Jules-A** [comment]
> I suggest that you open a new issue, as no one here knows what problems you're experiencing without filling out an issue template.

It's not so much a problem so much as it's a feature request. The feature I was after was pretty much exactly what you described in your OP of this issue.
Though I guess since this is closed, I guess https://github.com/mpv-player/mpv/issues/11122 is sort of similar though I've just after syncing to the highest integer multiple of video fps (though being able set a cap based on fps or max multiples would be nice). For example playing a 23.976hz video on a 165hz display would cap out at 143.856 fps.

**Comment by wiwaz** [comment]
What you're describing is not needed for displays with good VRR ranges or ones with LFC. display-resample is also completely incompatible with VRR as already stated. Stop wasting everyone's time with your unrelated problems.

**Comment by Jules-A** [comment]
> What you're describing is not needed for displays with good VRR ranges or ones with LFC. display-resample is also completely incompatible with VRR as already stated. Stop wasting everyone's time with your unrelated problems.

Sure but my monitor doesn't have LFC (48-165hz) and I'm pretty sure there's other people with the same problem. Also, doesn't LFC mess up interpolation?
How is it unrelated? My request is literally the same as what the original issue stated, it's really everyone else that's going off-topic...

**Comment by dechamps** [comment]
> I'm just after syncing to the highest integer multiple of video fps (though being able set a cap based on fps or max multiples would be nice). For example playing a 23.976hz video on a 165hz display would cap out at 143.856 fps.

Have you tried something along these lines?

```
video-sync=audio

[vrr1x]
profile-cond=get("container-fps", 999) >= 55
vf-remove=@vrr

[vrr2x]
profile-cond=get("container-fps", 999) < 55 and get("container-fps", 999) >= 35
vf-remove=@vrr
vf-add=@vrr:fps=fps=2*source_fps

[vrr3x]
profile-cond=get("container-fps", 999) < 35
vf-remove=@vrr
vf-add=@vrr:fps=fps=3*source_fps
```

**Comment by Jules-A** [comment]
> Have you tried something along these lines?

Wait, you can seriously remove vrr with vf-remove=@vrr ??? I haven't tried that.
However I did mess around with changing the source fps with vf but with my shaders it causes framedrops when doing >2x source. I did notice that with 2*source fps (which is below what my monitor's  VRR range) is at 2x the result of that (so 23.975 -> ~96) whereas without it, it's flickering between 68-78.

Also the main reason I don't settle with the vf is interpolation doesn't seem to work well with it.

**Comment by dechamps** [comment]
> > Have you tried something along these lines?
> 
> Wait, you can seriously remove vrr with `vf-remove=@vrr` ???

No. That's not what it does. The `vf-remove` statements are only there to clean up the fps filter when switching between the different profiles.
